### PR TITLE
feat: Support multiple WF status listeners via composite pattern

### PR DESCRIPTION
**Pull request by akhilpathivada**
Resolves https://github.com/conductor-oss/conductor/issues/727

### Summary

Implements support for multiple workflow status listeners using the Composite Pattern, enabling simultaneous publishing of workflow events to multiple destinations (Kafka, queues, webhooks, archive).

### Motivation

Previously, Conductor only allowed a single workflow status listener type to be configured via `conductor.workflow-status-listener.type`. Users had to choose between `kafka`, `queue_publisher`, `workflow_publisher`, or `archive` - but could not use multiple simultaneously.

This limitation prevented organizations from broadcasting the same workflow events to different systems for different purposes:

- Kafka: Real-time analytics and monitoring dashboards
- Queues: Internal workflow orchestration and cleanup jobs
- Webhooks: External system integrations (ticketing systems)

Each system has different requirements (stream processing vs guaranteed delivery vs fire-and-forget), different failure domains, and different retention policies.

### Changes

1. **CompositeWorkflowStatusListener**: Delegates workflow events to all configured listeners in parallel
2. **CompositeWorkflowStatusListenerConfiguration**: Spring configuration with factory method for dynamic listener creation
3. **CompositeWorkflowStatusListenerProperties**: Configuration properties for composite listener
4. Comprehensive unit tests with coverage for all workflow event types and failure scenarios

### Configuration Example
```
# Enable composite mode
conductor.workflow-status-listener.type=composite

# List the listeners to enable (comma-separated)
conductor.workflow-status-listener.composite.types=kafka,workflow_publisher,queue_publisher

# Each listener retains its existing configuration namespace

# Kafka configuration
conductor.workflow-status-listener.kafka.producer[bootstrap.servers]=kafka:29092
conductor.workflow-status-listener.kafka.default-topic=workflow-events

# Workflow publisher (webhook) configuration
conductor.status-notifier.notification.url=http://webhook-endpoint:8080/workflow-events

# Queue publisher configuration
conductor.workflow-status-listener.queue-publisher.successQueue=_callbackSuccessQueue
conductor.workflow-status-listener.queue-publisher.failureQueue=_callbackFailureQueue
```
### Testing
✅ Unit tests for all workflow event types 
✅ Failure isolation tests (single and multiple listener failures)
✅ Edge cases (empty list, single listener)
✅ Verification that all delegates are invoked correctly

### CODE REVIEW

**Comment by akhilpathivada** [comment]
I'll mark this as ready for review once https://github.com/conductor-oss/conductor/pull/720 is merged, as this PR is dependent on those changes.

**Comment by akhilpathivada** [comment]
@v1r3n could you please help reviewing this?

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    CompositeWorkflowStatusListenerProperties.class,
+    StatusNotifierNotificationProperties.class,
+    KafkaWorkflowStatusPublisherProperties.class,
+    ConductorQueueStatusPublisherProperties.class,
+    ArchivingWorkflowListenerProperties.class
+})
+@ConditionalOnProperty(name = "conductor.workflow-status-listener.type", havingValue = "composite")
+public class CompositeWorkflowStatusListenerConfiguration {
+
+    private static final Logger LOGGER =
```
Comment (nthmost-orkes) [inline workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java:48]: This `RestClientManager` bean is registered unconditionally whenever `type=composite` is active — even if `workflow_publisher` is not in the composite types list. This means a user who configures `types=kafka,queue_publisher` will still have `RestClientManager` instantiated, potentially failing at startup if `conductor.status-notifier.notification.url` is not configured.

This also makes the `Optional<RestClientManager>` parameter in `compositeWorkflowStatusListener` misleading, since the bean is always present.

The fix is to move `RestClientManager` construction inside the `workflow_publisher` case in `createListener()`, instead of registering it as a top-level bean.

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListener.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                workflow, "onWorkflowPaused", listener -> listener.onWorkflowPaused(workflow));
+    }
+
+    @Override
+    public void onWorkflowResumed(WorkflowModel workflow) {
+        delegateToListeners(
+                workflow, "onWorkflowResumed", listener -> listener.onWorkflowResumed(workflow));
+    }
+
+    @Override
```
Comment (nthmost-orkes) [inline workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListener.java:120]: `listeners.stream().parallel()` routes work through the JVM common ForkJoin pool, which is shared across the entire application. For the typical use case of 2–3 listeners, the overhead of the ForkJoin pool likely exceeds any parallelism benefit, and slow or blocking listeners could starve other pool users.

Consider either:
- Sequential iteration (`listeners.forEach(...)`) — simpler and sufficient when listeners are fast
- A dedicated `ExecutorService` with `CompletableFuture.allOf(...)` — if true parallel execution is important for latency

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                                            rcm,
+                                            executionDAOFacade,
+                                            notifierProperties
+                                                    .get()
+                                                    .getSubscribedWorkflowStatuses());
+                                })
+                        .orElseThrow(
+                                () ->
+                                        new IllegalStateException(
+                                                "Workflow publisher requested but notification properties not configured. "
```
Comment (nthmost-orkes) [inline workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java]: `notifierProperties.get()` will throw `NoSuchElementException` if the Optional is empty, bypassing the descriptive `orElseThrow` message below. Although `@EnableConfigurationProperties` makes this hard to trigger today, it is fragile. Consider `notifierProperties.map(StatusNotifierNotificationProperties::getSubscribedWorkflowStatuses).orElseThrow(...)` or restructuring to validate up-front.

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                                            props.getTtlDuration());
+
+                                    if (props.getTtlDuration().getSeconds() > 0) {
+                                        return new ArchivingWithTTLWorkflowStatusListener(
+                                                executionDAOFacade, props);
+                                    } else if (props.getWorkflowArchivalType()
+                                            == ArchivingWorkflowListenerProperties.ArchivalType
+                                                    .S3) {
+                                        return new ArchivingWorkflowToS3(executionDAOFacade, props);
+                                    } else {
```
Comment (nthmost-orkes) [inline workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java]: Returning `null` for an unknown type is silent degradation — a typo like `types=kafkaa` will result in that listener being silently dropped with only a WARN log, and no other indication of misconfiguration. Since this is evaluated at startup, consider throwing `IllegalArgumentException` here instead to fail fast and surface the error clearly.

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java (hunk 3 above)
Comment (akhilpathivada) [reply workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java]: Agree. I thought of `@EnableConfigurationProperties ` would always save from `notifierProperties.get()` being Empty - so I've written this way. I've restructured it now, according to the comments.

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListener.java (hunk 2 above)
Comment (akhilpathivada) [reply workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListener.java:120]: Fixed it. Went with option-1, since true parallel execution is not very critical for these kind of features.

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java (hunk 4 above)
Comment (akhilpathivada) [reply workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java]: Done. Thanks for calling this out - I've missed considering this, and throwing an expection!

File: workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java (hunk 1 above)
Comment (akhilpathivada) [reply workflow-event-listener/src/main/java/com/netflix/conductor/contribs/listener/composite/CompositeWorkflowStatusListenerConfiguration.java:48]: Done

**Comment by manan164** [comment]
Thanks for the contribution! A few things worth addressing before merging:

**Issue 1 — Sequential dispatch (docs say "parallel")**
The PR description and README say listeners run "in parallel" / "simultaneously", but the implementation uses `listeners.forEach(...)` which is strictly sequential. If Kafka is slow or blocked, every listener after it in the list is delayed by the same amount — worst case in exactly the failure scenario this feature is meant to isolate. Either switch to `CompletableFuture`-based parallel dispatch, or correct the docs to say "sequentially".

**Issue 2 — Configuration duplicates listener factory logic**
`CompositeWorkflowStatusListenerConfiguration` manually instantiates `KafkaWorkflowStatusPublisher`, `ConductorQueueStatusPublisher`, etc., duplicating logic that already lives in `KafkaWorkflowStatusPublisherConfiguration`, `ArchivingWorkflowListenerConfiguration`, and so on. This works today since constructor signatures match, but if a new dependency is added to any listener, it must be updated in two places. Consider delegating construction back to the existing config classes rather than reimplementing it here.

**Issue 3 — `Optional` injection doesn't guard what you think**
`CompositeWorkflowStatusListenerConfiguration` declares `@EnableConfigurationProperties` for all the properties types (Kafka, archive, queue, etc.), which means Spring always creates those beans when composite mode is active — the `Optional` parameters will never be `Optional.empty()`. The `orElseThrow` safety net is effectively dead code. In practice, if a user sets `types=kafka` but forgets to configure Kafka properties, they won't see your helpful error message — they'll get a cryptic failure from inside the Kafka producer with default/empty values. Consider validating that the required properties are actually populated (e.g., checking `bootstrapServers` is non-empty) rather than relying on `Optional` presence.
